dali/kernels/reduce/reduce_all_kernel_gpu.h

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_REDUCE_ALL_KERNEL_GPU_H_
+#define DALI_KERNELS_REDUCE_REDUCE_ALL_KERNEL_GPU_H_
+
+#include <memory>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/reduce/reduce.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce {
+
+template <typename T, int ndim, typename Reductor, typename Preprocessor = identity>
+class DLL_PUBLIC ReduceAllGPU {
+ public:
+  // TODO(janton): remove this when implemented
+  static_assert(std::is_same<Preprocessor, identity>::value,
+                "Preprocessing is not yet implemented");
+
+  DLL_PUBLIC ~ReduceAllGPU() = default;
+
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<T, ndim> &in) {
+    auto num_samples = in.size();
+    ScratchpadEstimator se;
+
+    int64_t max_sample_size = 0;
+    for (int i = 0; i < num_samples; i++) {
+      max_sample_size = std::max(max_sample_size, volume(in.tensor_shape(i)));
+    }
+    blocks_per_sample_ = std::max(32, static_cast<int>(std::sqrt(max_sample_size)));
```

Comment:
I'm starting to have second thoughts here. For small data (on ther order of < 4kB max) there's no point in splitting the reduction into two stages - the kernel invocation would constitute significant overhead. For such a case, you can force blocks_per_sample to 1 and make the first kernel produce the output directly.

Reply:
done